ユーザー管理機能

FjiiNamiki opened the pull request:
#What
Deviseを導入
ユーザー管理機能を実装
モデルの単体テストコードを書く

#Why
ユーザー管理機能を実装するため

- 正しく情報を記入すると、ユーザーの新規登録ができる動画
https://gyazo.com/b21177276bd3813a8cb1caad55f13bef

- 入力に問題のある状態ではユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/2d74bf64a763410c630eee1adf533209

- 正しく情報を記入すると、ユーザーのログインができる動画
https://gyazo.com/999c453ab6a7a2afb5a83b5467a5dc5b

- 入力に問題のある状態ではユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/04e63780eea5a98a60dece2499882965

- ログアウトに成功する動画
https://gyazo.com/40262dec2b55e9b5fc30a5b29fff5ced

- テスト成功の画像
https://gyazo.com/b7d9c096524d9c0fdd793d62035f34c0

FjiiNamiki committed:
Deviseの導入

FjiiNamiki committed:
バリデーションの設定

FjiiNamiki committed:
ログイン機能実装

FjiiNamiki committed:
ユーザー管理機能単体テストコードの実装

FjiiNamiki committed:
Rubocopの実行

koyamatakumi commented on app/models/user.rb:
```diff
@@ -0,0 +1,15 @@
+  validates :lastname_kana, presence: true, format: { with: /\A[ァ-ヶー]+\z/ }
+  validates :firstname_kana, presence: true, format: { with: /\A[ァ-ヶー]+\z/ }
+  validates :birthday, presence: true
+end
```
```  
validates :nickname, presence: true
  VALID_PASSWORD_REGEX = /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i
  validates :password, format: { with: VALID_PASSWORD_REGEX }
  validates :lastname_kanji, presence: true, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
  validates :firstname_kanji, presence: true, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
  validates :lastname_kana, presence: true, format: { with: /\A[ァ-ヶー]+\z/ }
  validates :firstname_kana, presence: true, format: { with: /\A[ァ-ヶー]+\z/ }
  validates :birthday, presence: true
```
全てにpresence: trueの記述があり冗長な記述かと存じます。
with_optionを用いてまとめて記述するとスッキリとした記述になるかと存じますので修正をお願いいたします！
ps.パスワードに関しては、デフォルトで設定されているため、with_option　presence: true　do ~endの外に記述しましょう！

koyamatakumi commented on app/views/shared/_header.html.erb:
```diff
@@ -16,10 +16,13 @@
+        <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
+      <% end %>
       <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
     </ul>
```
```
18行目 <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
26行目<%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
こちらは実装完了後は不要なコメントアウトかと存じます。消しておきましょう。

koyamatakumi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,116 @@
+        @user.password = 'aaaaaaaa'
+        @user.valid?
+        expect(@user.errors.full_messages).to include 'Password is invalid'
+      end
```
```
passwordが半角英数字混合での入力でないと登録できない
```
こちら英語のみの場合を記述していただいておりますので、同様に数字のみでも登録できないテストをしていただければと思います。

koyamatakumi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,116 @@
+        @user.firstname_kana = 'なみき'
+        @user.valid?
+        expect(@user.errors.full_messages).to include 'Firstname kana is invalid'
+      end
```
名字（カナ）と名前（カナ）は
それぞれ漢字の場合、英語の場合のテストを記述していただくと、より強固なテストになるかと存じますので、追記をお願いいたします。

koyamatakumi reviewed (commented):
レビュー依頼ありがとうございます！
4点コメントいたしましたので、ご確認をお願いいたします。
この後も実装頑張っていきましょう！

FjiiNamiki committed:
ユーザー管理機能修正1

バリデーションをwith_optionを用いてまとめて記述。
不要なコメントアウトを削除
パスワードのテストコードを数字のみも追加
名前（カナ）のテストコードを英字のみ、ひらがなのみ、漢字のみも追記

FjiiNamiki commented:
レビューありがとうございます。
修正しました。
# What
①バリデーションをwith_optionを用いてまとめて記述。
②不要なコメントアウトを削除
③パスワードのテストコードを数字のみも追加
④名前（カナ）のテストコードを英字のみ、ひらがなのみ、漢字のみも追記

# Why
①コードが冗長だったため
②消し忘れを削除
③より強固なテストにするため
④より強固なテストにするため

AsukaToyoshima commented on app/models/user.rb:
```diff
@@ -0,0 +1,19 @@
+
+  with_options presence: true do
+    validates :nickname
+    validates :lastname_kanji, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
+    validates :firstname_kanji, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
+    validates :lastname_kana, format: { with: /\A[ァ-ヶー]+\z/ }
+    validates :firstname_kana, format: { with: /\A[ァ-ヶー]+\z/ }
```
with_optionsを使用しまとめていただけましたね。
入れ子構造とすることで、さらに下記のようにまとめて記述することができます。
DRYを意識したコードとなり、可読性も向上するので実装してみましょう。
```Ruby
with_options presence: true do

   with_options ~~ do

    end

    with_options ~~ do

    end
end

AsukaToyoshima commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,140 @@
+      it '誕生日が空では登録できない' do
+        @user.birthday = ''
+        @user.valid?
+        # binding.pry
```
こちらのコメントにつきまして、不要であれば削除しておきましょう。

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
2点コメントさせていただきました。ご確認の上、修正をお願いいたします。

疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

FjiiNamiki committed:
ユーザー管理機能修正2

①バリデーションをwith_optionを用いてまとめて記述。
②不要なコメントアウトを削除

FjiiNamiki commented:
レビューありがとうございます。修正しました。

# What
①バリデーションをwith_optionを用いてまとめて記述。
②不要なコメントアウトを削除
Merge

# Why
①DRYを意識した可読性の高いコードにするため。
②消し忘れていたものを削除

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます。実装お疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！
